### PR TITLE
Clean up smoke test code and begin automating QA tests

### PR DESCRIPTION
**Pull request by squat**
These commits overhaul most of the smoke test code to simplify and
standardize the retry logic. This also lets us more easily extend the
tests to include QA tests, and eventually rewrite all bash scripts as
Go. Finally, it adds the first automated QA test: testing the status of the
stats-emitter.

cc @Quentin-M @rithujohn191 @kans 
Addresses: #1040

### CODE REVIEW

File: tests/smoke/smoke_test.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	t := time.NewTimer(d)
+	c := make(chan struct{})
+	go func() {
+		<-t.C
+		close(c)
+	}()
+	return c
+}
+
+func retry(r retriable, t *testing.T, period, limit time.Duration) error {
```
Comment (kans) [inline tests/smoke/smoke_test.go]: period/limit could have better names.

Comment (squat) [inline tests/smoke/smoke_test.go]: I agree. `interval` and `max`? Also, I'll rename all occurrences of `wait` to `max` or whatever

Comment (squat) [inline tests/smoke/smoke_test.go]: updated :)

File: tests/smoke/smoke_test.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	// kubeconfigEnv is the environment variable that is checked for a the kubeconfig path to be loaded.
+	kubeconfigEnv = "TEST_KUBECONFIG"
+)
+
+var (
+	runClusterTests bool
```
Comment (rithujohn191) [inline tests/smoke/smoke_test.go]: can we add some comments for these vars explaining what they do

Comment (squat) [inline tests/smoke/smoke_test.go]: Sure, I'll add that to the commits now. these two variables are flags used to gate white tests to run.

Comment (squat) [inline tests/smoke/smoke_test.go]: @rithujohn191 updated

**Comment by Quentin-M** [comment]
ok to test

**Comment by Quentin-M** [comment]
Let's run it a few more times.

**Comment by squat** [comment]
NAT gateway limit exceeded :/

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
Passed. Running again.

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
This test has passed like 5 times and has only failed due to resource limits. Running once more then merging.

**Comment by squat** [comment]
ok to test

**Comment by squat** [comment]
jenkins worker went offline. starting test again.

**Pull request merged**